python-package/xgboost/testing/data.py

```diff
@@ -124,3 +124,41 @@ def pd_dtypes() -> Generator:
         orig = pd.DataFrame(data, dtype=np.bool_ if Null is None else pd.BooleanDtype())
         df = pd.DataFrame(data, dtype=pd.BooleanDtype())
         yield orig, df
+
+
+def pd_arrow_dtypes() -> Generator:
+    """Pandas DataFrame with pyarrow backed type."""
+    import pandas as pd
+
+    import xgboost
+
+    # Integer
+    dtypes = xgboost.data.pandas_pyarrow_mapper
+    Null: Union[float, None, Any] = np.nan
+    orig = pd.DataFrame(
+        {"f0": [1, 2, Null, 3], "f1": [4, 3, Null, 1]}, dtype=np.float32
+    )
+    # Create a dictionary-backed dataframe, enable this when the roundtrip is
+    # implemented in pandas/pyarrow
+    # df = pd.DataFrame(
+    #     {"f0": [0, 2, Null, 3], "f1": [4, 3, Null, 1]},
+    #     dtype=pd.ArrowDtype(pa.dictionary(pa.int32(), pa.int32(), ordered=True)),
+    # )
+
+    for Null in (None, pd.NA):
+        for dtype in dtypes:
+            if dtype.startswith("float16") or dtype.startswith("bool"):
+                continue
+            df = pd.DataFrame(
+                {"f0": [1, 2, Null, 3], "f1": [4, 3, Null, 1]}, dtype=dtype
+            )
+            yield orig, df
+
+    # Create a boolean array, error:
+    # Could not convert <pyarrow.NullScalar: None> with type pyarrow.lib.NullScalar:
+    # tried to convert to boolean
+    #
+    # df = pd.DataFrame(
+    #     {"f0": [True, False, pa.NA, True], "f1": [False, True, pa.NA, True]},
```

Comment:
Not sure if it was intentional to test p**a**.NA instead of p**d**.NA, but generally pandas doesn't support constructing with pyarrow scalars.

Also doesn't seem like pyarrow supports this either

```
In [8]: pa.array([True, pa.NA])
---------------------------------------------------------------------------
ArrowInvalid                              Traceback (most recent call last)
Cell In [8], line 1
----> 1 pa.array([True, pa.NA])

File ~/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/pyarrow/array.pxi:317, in pyarrow.lib.array()

File ~/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/pyarrow/array.pxi:39, in pyarrow.lib._sequence_to_array()

File ~/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/pyarrow/error.pxi:144, in pyarrow.lib.pyarrow_internal_check_status()

File ~/opt/miniconda3/envs/pandas-dev/lib/python3.8/site-packages/pyarrow/error.pxi:100, in pyarrow.lib.check_status()

ArrowInvalid: Could not convert <pyarrow.NullScalar: None> with type pyarrow.lib.NullScalar: tried to convert to boolean
```

Reply:
Thank you for sharing! Changed it to pandas NA and enabled the test for boolean. I have also added some more notes on the use of dictionary. ;-)